fluent.syntax/fluent/syntax/ast.py

```diff
@@ -3,6 +3,35 @@
 import json
 
 
+class Visitor(object):
+    '''Read-only visitor pattern.
+
+    Subclass this to gather information from an AST.
+    To generally define which nodes not to descend in to, overload
+    `generic_visit`.
+    To handle specific node types, add methods like `visit_Pattern`.
+    The boolean value of the returned value determines if the visitor
+    descends into the children of the given AST node.
+    '''
+    def visit(self, value):
+        if isinstance(value, BaseNode):
+            self.visit_node(value)
+        if isinstance(value, list):
+            for node in value:
+                self.visit(node)
+
+    def visit_node(self, node):
+        nodename = type(node).__name__
+        entering = getattr(self, 'visit_{}'.format(nodename), self.generic_visit)
+        if not entering(node):
+            return
+        for propname, propvalue in vars(node).items():
+            self.visit(propvalue)
```

Comment:
I'd like to suggest a slightly more verbose way of writing this method. It's a nit, so feel free to ignore. As you know, I've been trying to promote the idea of self-documenting code in my reviews :) Here, the value I think this brings is that it gives an explicit meaning to the return value of the `process_node` method.

```
def visit_node(self, node):
        nodename = type(node).__name__
        process_node = getattr(self, 'visit_{}'.format(nodename), self.generic_visit)
        should_continue = process_node(node)
        if should_continue:
            for propname, propvalue in vars(node).items():
                self.visit(propvalue)
```